### PR TITLE
nanocoap: add serverside block1 support

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR makes use of the new option handling (introduced in #8772) and adds support for server-side blockwise (block1) support.

### Issues/PRs references

~~Waiting for #8772.~~

### CODE REVIEW

**Comment by kaspar030** [comment]
- rebased after #8772 merge, ready for review

File: sys/include/net/nanocoap.h [hunk 1]
```diff
@@ -436,6 +516,18 @@ unsigned coap_get_content_type(coap_pkt_t *pkt);
  */
 int coap_get_uri(coap_pkt_t *pkt, uint8_t *target);
 
+/**
+ * @brief    Helper to convert SZX value to size
+ *
+ * @param[in]   szx     SZX value to convert
+ *
+ * @returns     size
```
Comment (bergzand) [inline sys/include/net/nanocoap.h]: Maybe expand a bit to "size in bytes" to clarify a bit.

**Comment by bergzand** [comment]
I'm doing some extensive testing (feeding gigabyte size data to the sha256 tester) and I'm seeing some weird things:
```
_sha256_handler(): received data: offset=6080 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=6144 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=6208 len=65440 blockwise=1 more=1
_sha256_handler(): received data: offset=6272 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=6336 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=6400 len=65440 blockwise=1 more=1
_sha256_handler(): received data: offset=6464 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=6528 len=64 blockwise=1 more=1
```
snip
```
_sha256_handler(): received data: offset=52160 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=52224 len=26111 blockwise=1 more=1
_sha256_handler(): received data: offset=52288 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=52352 len=26111 blockwise=1 more=1
_sha256_handler(): received data: offset=52416 len=64 blockwise=1 more=1
_sha256_handler(): received data: offset=52480 len=26111 blockwise=1 more=1
_sha256_handler(): received data: offset=52544 len=64 blockwise=1 more=1
```

This is produced with the nanocoap_server example on native using libcoaps coap-client to feed it data:
`coap-client coap://[fe80::c0dd:c1ff:fe29:833b%tapbr0]/sha256 -m POST -f ~/$LARGE_FILE  -b 64`

I've wiresharked the packets, they all have size 64 there.

File: sys/include/net/nanocoap.h [hunk 2]
```diff
@@ -409,6 +432,63 @@ size_t coap_put_option_ct(uint8_t *buf, uint16_t lastonum, uint16_t content_type
  */
 size_t coap_put_option_uri(uint8_t *buf, uint16_t lastonum, const char *uri, uint16_t optnum);
 
+/**
+ * @brief    Generic block option getter
+ *
+ * @param[in]   pkt     pkt to work on
+ * @param[in]   option  actual block option number to get
+ * @param[out]  blknum  block number
+ * @param[out]  szx     SZX value
```
Comment (bergzand) [inline sys/include/net/nanocoap.h]: Maybe clarify here that these parameters are modified to zero if the option is not found.

File: examples/nanocoap_server/coap_handler.c [hunk 3]
```diff
@@ -55,9 +56,58 @@ static ssize_t _riot_value_handler(coap_pkt_t *pkt, uint8_t *buf, size_t len, vo
...
+        pkt_pos += fmt_bytes_hex((char *)pkt_pos, digest, sizeof(digest));
+    }
+
+    return pkt_pos - (uint8_t*)pkt->hdr;
+}
+
 /* must be sorted by path (alphabetically) */
 const coap_resource_t coap_resources[] = {
     COAP_WELL_KNOWN_CORE_DEFAULT_HANDLER,
+    { "/sha256", COAP_POST | COAP_PUT, _sha256_handler, NULL },
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply examples/nanocoap_server/coap_handler.c]: fixed

File: sys/include/net/nanocoap.h [hunk 4]
```diff
@@ -409,6 +432,63 @@ size_t coap_put_option_ct(uint8_t *buf, uint16_t lastonum, uint16_t content_type
...
+ *
+ * @returns     -1 if option not found
+ * @returns     0 if more flag is not set
+ * @returns     1 if more flag is set
+ */
+int coap_get_blockopt(coap_pkt_t *pkt, uint16_t option, uint32_t *blknum, unsigned *szx);
+
+/**
+ * @brief    Block1 helper
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply sys/include/net/nanocoap.h]: ok

**Comment by kaspar030** [comment]
> I'm doing some extensive testing (feeding gigabyte size data to the sha256 tester) and I'm seeing some weird things:

Does this happen a lot? I've been feeding q couple hundred jiggabytes through the tester and so far, I cannot reproduce.

File: sys/include/net/nanocoap.h [hunk 5]
```diff
@@ -409,6 +432,63 @@ size_t coap_put_option_ct(uint8_t *buf, uint16_t lastonum, uint16_t content_type
...
+ * @param[in]   pkt     pkt to work on
+ * @param[out]  block1  ptr to preallocated coap_block1_t structure
+ *
+ * @returns     0 if block1 option not present
+ * @returns     1 if structure has been filles
+ */
+int coap_get_block1(coap_pkt_t *pkt, coap_block1_t *block1);
+
+/**
+ * @brief   Insert block1 option into buffer
```
[the remark this answers is not in the document]
Comment (bergzand) [reply sys/include/net/nanocoap.h]: Ok

**Comment by bergzand** [comment]
> Does this happen a lot? I've been feeding q couple hundred jiggabytes through the tester and so far, I cannot reproduce.

After testing a bit more this appears to be a variant of #6123, except that I'm hammering the instance here with CoAP packets instead of ping floods. This results in sock returning a far too large packet size (several times more than the mtu size). So this issue is unrelated to this PR.

**Comment by bergzand** [comment]
@kaspar030 It looks like the commits are in the wrong order, could you rebase it to first the block1 support and then the commit adding example. Otherwise people are going to complain when bisecting.

**Comment by kaspar030** [comment]
@bergzand github often messes up the order. the actual commit order in the branch is as you suggested.

**Comment by bergzand** [comment]
Ah, you're right, never mind then

**Pull request merged**